objectbox/lib/src/store.dart

```diff
@@ -93,35 +101,66 @@ class Store {
   ///
   /// Don't try to call any other ObjectBox methods after the store is closed.
   void close() {
-    // Call each "onBeforeClose()" event listener.
+    _boxes.values.forEach(InternalBoxAccess.close);
+    _boxes.clear();
+
+    // Call each "onClose()" event listener.
     // Move the list to prevent "Concurrent modification during iteration".
-    final listeners = StoreCloseObserver.removeAllListeners(this);
-    listeners.forEach((listener) => listener());
+    _onClose.values.toList(growable: false).forEach((listener) => listener());
+    _onClose.clear();
 
     checkObx(C.store_close(_cStore));
   }
 
   /// Returns a cached Box instance.
-  Box<T> box<T>() => Box<T>(this);
+  Box<T> box<T>() {
+    if (!_boxes.containsKey(T)) {
+      return _boxes[T] = InternalBoxAccess.create<T>(this, _entityDef());
+    }
+    return _boxes[T] as Box<T>;
+  }
 
-  EntityDefinition<T> entityDef<T>() {
-    final binding = defs.bindings[T];
+  EntityDefinition<T> _entityDef<T>() {
+    final binding = _defs.bindings[T];
     if (binding == null) {
       throw ArgumentError('Unknown entity type ' + T.toString());
     }
     return binding /*!*/ as EntityDefinition<T>;
   }
 
-  /// Executes a given function inside a transaction.
+  /// Executes a given function inside a transaction. Returns [fn]'s result.
+  /// Aborts a transaction or rethrows if there's an exception.
   ///
-  /// Returns type of [fn] if [return] is called in [fn].
+  /// A transaction can group several operations into a single unit of work that
+  /// either executes completely or not at all.
+  /// The advantage of explicit transactions over the bulk put operations is
+  /// that you can perform any number of operations and use objects of multiple
+  /// boxes. In addition, you get a consistent (transactional) view on your data
+  /// while the transaction is in progress.
   R runInTransaction<R>(TxMode mode, R Function() fn) =>
       Transaction.execute(this, mode, fn);
 
-  /// Return an existing SyncClient associated with the store or null if not available.
-  /// See [Sync.client()] to create one first.
+  /// Return an existing SyncClient associated with the store or null if not
+  /// available. Use [Sync.client()] to create one first.
   SyncClient /*?*/ syncClient() => syncClientsStorage[this];
 
   /// The low-level pointer to this store.
   Pointer<OBX_store> get ptr => _cStore;
 }
+
+/// Internal only.
+// TODO enable annotation once meta:1.3.0 is out
+// @internal
+class InternalStoreAccess {
```

Comment:
These internal access classes still make me cringe. Shouldn't public interfaces and internal implementations hide API from users?

Reply:
Yeah, dislike it too, but couldn't think of a better way to have some "package-protected" access among `lib/src/*.dart` files. The problem is, once some identifier starts with an underscore, you can only access it from the same file, because in Dart, file is a library. Of course you can also use `part` instead of import, basically doing the same thing - turning everything into a single library...

Reply:
As I said, isn't that what interfaces are for? E.g. make `Box` an abstract class (as Dart apparently does not have explicit interfaces) and create `BoxImpl` as the actual implementation. The implementation can have any number of public methods that users (which work with the interface) do not see (unless they use the implementation class directly, which docs should discourage).